[lldb] Print a message when background tasks take a while to complete

JDevlieghere opened the pull request:
When terminating the debugger, we wait for all background tasks to
complete. Given that there's no way to interrupt those treads, this can
take a while. When that happens, the debugger appears to hang at exit.

The above situation is unfortunately not uncommon when background
downloading of dSYMs is enabled (`symbols.auto-download background`).
Even when calling dsymForUUID with a reasonable timeout, it can take a
while to complete.

This patch improves the user experience by printing a message from the
driver when it takes more than one (1) second to terminate the debugger.

clayborg commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -623,8 +630,20 @@ void Debugger::Terminate() {
   if (g_thread_pool) {
-    // The destructor will wait for all the threads to complete.
-    delete g_thread_pool;
+    // Delete the thread pool in a different thread so we can set a timeout.
```
Can we create a Progress object here? Maybe something like:
```
Progress progress("Shutting down worker threads");
```
Or is it too late and no one would receive events?

Another option would be to have the thread pool detach from all threads so that we don't need to wait until they complete and just go ahead and exit?

JDevlieghere replied:
> Or is it too late and no one would receive events?

Yup, that's exactly the problem. The debugger has already been destroyed at this point, otherwise I could've written to the debugger's output stream. By the time we terminate, there's pretty much nothing left at this point.   Before I had the callback I just wrote to stderr directly, but that's not a very nice thing to do as a library. 

> Another option would be to have the thread pool detach from all threads so that we don't need to wait until they complete and just go ahead and exit?

That's what we did in the past and that caused a bunch of crashes. If you want to do that without crashing, you need to be very careful about what you do in a task running on the thread pool, because the debugger might have been destroyed and terminated while you were running.

clayborg reviewed (commented):
Can we not just use a `Progress` object instead of installing a callback? Seems like a very specific use case to install a callback for in our public API. Do we create `Progress` objects for each background download? If we don't, should we? Then this would show up in the command line like "Downloading symbols for ..." and it might communicate to the user what is happening better than "Waiting for background tasks to complete...".

What would happen if we could add an internal API that would cancel all background downloads of symbols?

I am not sure I like installing a specific callback for thread pool issues. What if a thread gets deadlocked waiting for data and it never completes the download?

JDevlieghere commented:
> Can we not just use a `Progress` object instead of installing a callback? Seems like a very specific use case to install a callback for in our public API. Do we create `Progress` objects for each background download? If we don't, should we? Then this would show up in the command line like "Downloading symbols for ..." and it might communicate to the user what is happening better than "Waiting for background tasks to complete...".

We do emit progress events for all the downloads, but there's two reasons this doesn't work as you expect:

1. Currently, the default event handler ignores events that are shadowed. So if I kick off 4 downloads at the same time, we'll show the progress event for the first one only. The reason for this how we use ANSI escape codes to redraw the current line. I actually have something in the works to change that, but even if I can fix that, there's a second issue:

2. The thread pool is global and by the time we terminate, the debugger has already been destroyed, which includes it event handler thread and even its output stream. 

> What would happen if we could add an internal API that would cancel all background downloads of symbols?

That's what I described at the bottom of the description. You could come up with an ad-hoc solution for symbol downloads, but if we want to lean on the thread pool more, we'll see the same issue in other places too. If we want to go that route I think we should do it at the ThreadPool level and make all tasks (cooperatively) interruptible.  

> I am not sure I like installing a specific callback for thread pool issues. What if a thread gets deadlocked waiting for data and it never completes the download?

That's already a risk. The [ThreadPool](https://llvm.org/doxygen/classllvm_1_1ThreadPool.html) explicitly provides no guarantees around deadlocking.

JDevlieghere commented:
Greg and I talked this over offline. We both feel that progress events are the best way to convey that work is going on in the background. To make that work for the driver, we'll check if the debugger being destroyed is the very last one and if it is, we'll delay its destruction until the thread pool as completed. That still leaves the first problem of not shadowing progress events, but that can be deal with separately.

jasonmolenda reviewed (approved):
I haven't used the std::async feature before, but it looks reasonable.  This looks good to me.

JDevlieghere commented:
I tried implementing this and waiting on the thread pool when the last debugger is being destroyed is easy:

```
  // Only hold the debugger list lock long enough to check if we're the last
  // debugger being destroyed.
  bool last_debugger_being_destroyed = false;
  if (g_debugger_list_ptr && g_debugger_list_mutex_ptr) {
    std::lock_guard<std::recursive_mutex> guard(*g_debugger_list_mutex_ptr);
    last_debugger_being_destroyed =
        g_debugger_list_ptr->size() == 1 &&
        (*g_debugger_list_ptr)[0].get() == debugger_sp.get();
  }

  // Check if we should wait on the thread pool. If the debugger being destroyed
  // is the very last one, we can safely assume work is being done on its
  // behalf. We want to keep the debugger alive at least until the background
  // tasks are finished so they can report their progress.
  if (last_debugger_being_destroyed)
    g_thread_pool->wait();
```

However, at that point there's no default event handler thread anymore: we tear it down when exiting `RunCommandInterpreter`. I think I'm just going to move the timeout to the driver and call it a day...

JDevlieghere committed:
[lldb] Print a message when background tasks take a while to complete

When terminating the debugger, we wait for all background tasks to
complete. Given that there's no way to interrupt those treads, this can
take a while. When that happens, the debugger appears to hang at exit.

The above situation is unfortunately not uncommon when background
downloading of dSYMs is enabled (`symbols.auto-download background`).
Even when calling dsymForUUID with a reasonable timeout, it can take a
while to complete.

This patch improves the user experience by printing a message from the
driver when it takes more than one (1) second to terminate the debugger.
